src/cargo/core/resolver/types.rs

```diff
@@ -176,6 +176,14 @@ impl ActivationsKey {
     }
 }
 
+impl std::hash::Hash for ActivationsKey {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.0.fast_hash(state);
+        self.1.hash(state);
+        // self.2.hash(state); // Packages that only differ by SourceId are rare enough to not be worth hashing
```

Comment:
So we have 3 improvments
- Re-order
- Pointer hash on the name
- Skipping the source hash

Do we know how much of an improvement each one of these is?  I'm mostly wondering about skipping the source hash as that feels icky for some reason.

Not going to give this feeling too much weight though.  I'll mostly defer to you on this part.

Reply:
> Do we know how much of an improvement each one of these is?

I just collected overall performance numbers, I will try and get you some intermediate numbers.

>  I'm mostly wondering about skipping the source hash as that feels icky for some reason.

Can you articulate why? HashMaps can  devolve to `O(n)`  if  skip fields are the only thing distinguishing `n`  items, but we specifically commented to record why we didn't think that would happen. Skipping a field when hashing is a pretty standard performance trick.